### DISCUSSION TITLE
Zustand injectors

### DISCUSSION THREAD

**Discussion by nixjs** [pmndrs/zustand#2129] [Ideas] [answered]
I work with redux a lot, my redux store is very large. Therefore I used the redux-injectors library.
It allows to automatically load reducers if needed, instead of loading them all first. It helps increase web performance.
I zustand should also have the feature to automatically download the store if needed.
**Reference:** 
https://www.npmjs.com/package/redux-injectors
<img width="738" alt="image" src="https://github.com/pmndrs/zustand/assets/23132483/6b7f320a-4296-4295-bb8d-e4abb82d9364">

**Comment by dbritto-dev** [comment]
Nice, I'll take a look of that

**Reply by dbritto-dev** [reply]
@nixjs what would be the goal here? Having a way to lazy inject/load slices?

**Reply by nixjs** [reply]
This is a type of lazy loading, instead of loading once when entering the website.

**Reply by dbritto-dev** [reply]
@nixjs yeah I know what's lazy loading but in the case of zustand we don't have reducer and there's two options if you want to split a store: multiple stores or multiple slices

So lazy loading multiple stores doesn't make sense here but lazy loading slices would make sense

**Reply by thaind97git** [reply]
Hi @dbritto-dev , how to lazy loading slices? Can you give me an example?

**Reply by dbritto-dev** [reply]
@thaind97git here you go -> https://codesandbox.io/p/sandbox/prod-river-76fs6f?file=%2Fsrc%2FApp.tsx%3A1%2C10

**Reply by eltaljohn** [reply]
@dbritto-dev the project example code does not contain any Zustand code. Has it been restored of something like that?

**Reply by dbritto-dev** [reply]
@eltaljohn yeah, seems it was restored to the basic example

**Reply by eltaljohn** [reply]
@dbritto-dev could you put the example back to review the code-splitting?, please.

**Reply by dbritto-dev** [reply]
@eltaljohn sure I can create a new example

**Reply by eltaljohn** [reply]
Hi @dbritto-dev, Do you have any update on this?

**Reply by dbritto-dev** [reply]
@eltaljohn  -> https://codesandbox.io/p/devbox/lazy-load-slices-76fs6f

**Reply by eltaljohn** [reply]
I'm getting this error when enter the link:

![Screenshot 2024-02-15 at 1 59 56 PM](https://github.com/pmndrs/zustand/assets/35228599/33880ef4-f384-44fa-ba83-2a123ea02d11)

**Reply by dbritto-dev** [reply]
I guess I need to create a new sandbox on stackblitz, seems codesandbox change their policies and now I ran out of credits

**Comment by HousseinSA** [comment]
What to thank you guys for create this amazing state manager

**Comment by akhilshastri** [comment]
sorry, may be i m not getting it correctly, can we have code split with zustand ..? pls share the doc link

**Reply by dbritto-dev** [reply]
@akhilshastri yes, you can. we don't have a guide for that but I could add it.

**Reply by wcwts0** [reply]
Hi @dbritto-dev, when will the guide document be ready? Or are there any new stackblitz demos? Thanks.

**Reply by dbritto-dev** [reply, marked as answer]
@wcwts0 I'm working on package to add support for zustand injectors as an external package, the only issue that I faced now is the typings haha, but should be available as `zustand-injectors` so you can use it like another middleware for zustand

**Reply by wcwts0** [reply]
@dbritto-dev Great job! Looking forward to the package.

**Reply by dbritto-dev** [reply]
@nixjs @wcwts0 @akhilshastri @eltaljohn good news -> 

![image](https://github.com/pmndrs/zustand/assets/1697714/2b26064d-b6b8-4191-8aaf-afba10bdf45b)

**Reply by dbritto-dev** [reply]
```ts
import { createStore } from "zustand/vanilla";
import { describe, expect, test, vi } from "vitest";
import { injectors } from "../src";

describe("injectors", () => {
  test("should auto inject slices", async () => {
    const incrementBears = () => {};
    const store = createStore<{ bears: number; incrementBears: () => void }>()(
      injectors(
        () => ({
          bears: 0,
          incrementBears,
        }),
        {
          "increment-bears": async () =>
            new Promise((resolve) => {
              setTimeout(
                () =>
                  resolve((set) => ({
                    incrementBears: () => {
                      set((currentState) => ({
                        bears: currentState.bears + 1,
                      }));
                    },
                  })),
                300,
              );
            }),
        },
      ),
    );

    expect(store.getState().incrementBears).toBe(incrementBears);

    await vi.waitFor(() => {
      expect(store.getState().incrementBears).not.toBe(incrementBears);
    });
  });

  test("should manual inject slices", async () => {
    const incrementBears = () => {};
    const store = createStore<{ bears: number; incrementBears: () => void }>()(
      injectors(() => ({
        bears: 0,
        incrementBears,
      })),
    );

    expect(store.getState().incrementBears).toBe(incrementBears);

    store.injectAsyncSliceInitializer(
      "increment-bears",
      async () =>
        new Promise((resolve) => {
          setTimeout(
            () =>
              resolve((set) => ({
                incrementBears: () => {
                  set((currentState) => ({
                    bears: currentState.bears + 1,
                  }));
                },
              })),
            300,
          );
        }),
    );

    await vi.waitFor(() => {
      expect(store.getState().incrementBears).not.toBe(incrementBears);
    });
  });

  test("should actions works from injected slices", async () => {
    const store = createStore<{ bears: number; incrementBears: () => void }>()(
      injectors(() => ({
        bears: 0,
        incrementBears: () => {},
      })),
    );

    expect(store.getState().bears).toBe(0);

    store.getState().incrementBears();

    expect(store.getState().bears).toBe(0);

    store.injectAsyncSliceInitializer(
      "increment-bears",
      async () =>
        new Promise((resolve) => {
          setTimeout(
            () =>
              resolve((set) => ({
                incrementBears: () => {
                  set((currentState) => ({
                    bears: currentState.bears + 1,
                  }));
                },
              })),
            300,
          );
        }),
    );

    await vi.waitFor(() => {
      store.getState().incrementBears();

      expect(store.getState().bears).toBe(1);
    });
  });
});
```

**Reply by dbritto-dev** [reply]
@nixjs @wcwts0 @akhilshastri @eltaljohn here you go -> https://github.com/zustandjs/zustand-injectors
